File: python/src/pi_mono/coding_agent/core/source_info.py
"""Source metadata for resources and extensions."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, TypedDict

SourceScope = Literal["user", "project", "temporary"]
SourceOrigin = Literal["package", "top-level"]


class PathMetadata(TypedDict, total=False):
    source: str
    scope: SourceScope
    origin: SourceOrigin
    baseDir: str


@dataclass(frozen=True)
class SourceInfo:
    path: str
    source: str
    scope: SourceScope
    origin: SourceOrigin
    base_dir: str | None = None


def create_source_info(path: str, metadata: PathMetadata) -> SourceInfo:
    return SourceInfo(
        path=path,
        source=metadata["source"],
        scope=metadata.get("scope", "temporary"),
        origin=metadata.get("origin", "top-level"),
        base_dir=metadata.get("baseDir"),
    )


def create_synthetic_source_info(
    path: str,
    *,
    source: str,
    scope: SourceScope = "temporary",
    origin: SourceOrigin = "top-level",
    base_dir: str | None = None,
) -> SourceInfo:
    return SourceInfo(
        path=path,
        source=source,
        scope=scope,
        origin=origin,
        base_dir=base_dir,
    )

File: packages/coding-agent/src/core/source-info.ts
import type { PathMetadata } from "./package-manager.ts";

export type SourceScope = "user" | "project" | "temporary";
export type SourceOrigin = "package" | "top-level";

export interface SourceInfo {
	path: string;
	source: string;
	scope: SourceScope;
	origin: SourceOrigin;
	baseDir?: string;
	changelogPath?: string;
}

export interface SourcePackageIdentity {
	source: string;
	scope: SourceScope;
}

export function getSourcePackageKey(identity: SourcePackageIdentity): string {
	return JSON.stringify({
		source: identity.source,
		scope: identity.scope,
	});
}

export function createSourceInfo(path: string, metadata: PathMetadata): SourceInfo {
	return {
		path,
		source: metadata.source,
		scope: metadata.scope,
		origin: metadata.origin,
		baseDir: metadata.baseDir,
		...(metadata.changelogPath ? { changelogPath: metadata.changelogPath } : {}),
	};
}

export function createSyntheticSourceInfo(
	path: string,
	options: {
		source: string;
		scope?: SourceScope;
		origin?: SourceOrigin;
		baseDir?: string;
		changelogPath?: string;
	},
): SourceInfo {
	return {
		path,
		source: options.source,
		scope: options.scope ?? "temporary",
		origin: options.origin ?? "top-level",
		baseDir: options.baseDir,
		...(options.changelogPath ? { changelogPath: options.changelogPath } : {}),
	};
}
